allennlp/common/file_utils.py

```diff
@@ -21,37 +22,42 @@
 
 def url_to_filename(url: str, etag: str = None) -> str:
     """
-    Converts a url into a filename in a reversible way.
-    If `etag` is specified, add it on the end, separated by a period
-    (which necessarily won't appear in the base64-encoded filename).
-    Get rid of the quotes in the etag, since Windows doesn't like them.
+    Convert `url` into a hashed filename in a repeatable way.
+    If `etag` is specified, append its hash to the url's, delimited
+    by a period.
     """
     url_bytes = url.encode('utf-8')
-    b64_bytes = base64.b64encode(url_bytes)
-    decoded = b64_bytes.decode('utf-8')
+    url_hash = sha256(url_bytes)
+    filename = url_hash.hexdigest()
 
     if etag:
-        # Remove quotes from etag
-        etag = etag.replace('"', '')
-        return f"{decoded}.{etag}"
-    else:
-        return decoded
+        etag_bytes = etag.encode('utf-8')
+        etag_hash = sha256(etag_bytes)
+        filename += '.' + etag_hash.hexdigest()
+
+    return filename
 
-def filename_to_url(filename: str) -> Tuple[str, str]:
+def filename_to_url(filename: str, cache_dir: str = None) -> Tuple[str, str]:
     """
-    Recovers the the url from the encoded filename. Returns it and the ETag
-    (which may be ``None``)
+    Return the url and etag (which may be ``None``) stored for `filename`.
+    Raise ``FileNotFoundError`` if `filename` or its stored metadata do not exist.
     """
-    try:
-        # If there is an etag, it's everything after the first period
-        decoded, etag = filename.split(".", 1)
-    except ValueError:
-        # Otherwise, use None
-        decoded, etag = filename, None
+    if cache_dir is None:
+        cache_dir = DATASET_CACHE
 
-    filename_bytes = decoded.encode('utf-8')
-    url_bytes = base64.b64decode(filename_bytes)
-    return url_bytes.decode('utf-8'), etag
+    cache_path = os.path.join(cache_dir, filename)
+    if not os.path.exists(cache_path):
+        raise FileNotFoundError("file {} not found".format(cache_path))
+
+    meta_path = cache_path + '.json'
+    if not os.path.exists(meta_path):
+        raise FileNotFoundError("file {} not found".format(meta_path))
+
+    metadata = json.load(open(meta_path))
```

Comment:
`with` block? 😇

Reply:
It's not necessary. I'm pretty sure file objects close when they are out of scope, which here is as soon as the `json.load()` function call finishes. But it's no problem to use a `with` block. Explicit (or slightly less implicit) is better than implicit, after all.